tests/matchers.py

```diff
@@ -44,6 +44,39 @@ def __repr__(self):
         return repr_(self, self.repr_attrs)
 
 
+class InstanceOf(Matcher):
+    """Matches any instance of the given class with the given attrs.
+
+    As with Python's builtin isinstance() `class_` can be either a single class
+    or a tuple of classes (in which case the matcher will match instances of
+    *any* of the classes in the tuple).
+
+    If no `with_attrs` argument is given then the matcher will match instances
+    of the given class(es) without checking any attributes.
+    """
+
+    def __init__(self, class_, with_attrs=None):
+        self.class_ = class_
+        self.with_attrs = with_attrs or {}
+
+    @property
+    def repr_attrs(self):
+        if self.with_attrs:
+            return ("class_", "with_attrs")
+
+        return ("class_",)
+
+    def __eq__(self, other):
+        if not isinstance(other, self.class_):
+            return False
+
+        for name, value in self.with_attrs.items():
+            if not getattr(other, name) == value:
+                return False
+
+        return True
```

Comment:
This replaces h-matchers's `Any.instance_of()`:

```python
Any.instance_of(Foo)
matchers.InstanceOf(Foo)

Any.instance_of(Foo).with_attrs({"bar": "BAR", "gar": "GAR"})
matchers.InstanceOf(Foo, with_attrs={"bar": "BAR", "gar": "GAR"})
```

Also replaces [the pre-h-matchers `InstanceOf` matcher](https://github.com/hypothesis/h/blob/029a6c7a6aefd0388014dc9e6a9e349e3e8589e9/tests/common/matchers.py#L50-L60).

We could probably make it work like `matchers.InstanceOf(...).with_attrs({...})` fairly easily but I'm not sure it's obviously better.

Other stylistic tweaks we could make could be renaming the `matchers` fixture to `any` or `Any` and changing the `InstanceOf` naming style to `instance_of` 🤷‍♂️